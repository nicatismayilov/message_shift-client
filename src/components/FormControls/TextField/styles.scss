.text-field {
  position: relative;
  width: 100%;
  margin-bottom: 25px;
  font-family: "Quicksand", sans-serif;

  &__input {
    width: 100%;
    padding: 10px 16px;
    color: #2b2c43;
    font-size: 16px;
    font-weight: 500;

    background-color: #fff;
    border: 1px solid #e7e7e7;
    border-radius: 8px;
    transition: all 200ms;

    z-index: 2;

    &:focus {
      outline: none;
      border-color: #c9c7c7;
    }

    &--error {
      border-color: #ff5252;

      &:focus {
        border-color: #ff5252;
      }
    }
  }

  &__input--error + &__label {
    color: #ff5252;
  }

  &__label {
    position: absolute;
    top: 50%;
    left: 16px;
    transform: translateY(-50%);
    color: #2b2c43;
    font-size: 16px;
    font-weight: 600;
    transition: all 200ms ease-out;

    &--active {
      top: 0;
      left: 10px;
      font-size: 12px;
      padding: 0 2px;
      background-color: #fff;
      border-radius: 150px;
    }
  }

  &__hint {
    font-size: 12px !important;
    font-weight: 600 !important;

    position: absolute;
    left: 5%;
    top: 90%;
    z-index: 1;
    visibility: hidden;
    opacity: 0;

    transition: all 150ms;

    &--visible {
      top: 105%;
      visibility: visible;
      opacity: 1;
    }

    &--error {
      color: #ff5252;
    }
  }

  &__input:focus + &__label {
    top: 0;
    left: 10px;
    font-size: 12px;
    padding: 0 2px;
    background-color: #fff;
    border-radius: 150px;
  }
}
