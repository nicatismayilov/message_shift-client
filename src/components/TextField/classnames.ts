export const inputClassFunc = (hasError: boolean) => ({
	"text-field__input": true,
	"text-field__input--error": hasError,
});

export const labelClassFunc = (hasValue: boolean) => ({
	"text-field__label": true,
	"text-field__label--active": hasValue,
});

export const hintClassFunc = (isHintVisible: boolean, hasError: boolean) => ({
	"text-field__hint": true,
	"text-field__hint--visible": isHintVisible,
	"text-field__hint--error": hasError,
});
