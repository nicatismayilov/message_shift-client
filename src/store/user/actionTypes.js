export const AUHTENTICATE_USER_START = "AUHTENTICATE_USER_START";
export const AUHTENTICATE_USER_SUCCESS = "AUHTENTICATE_USER_SUCCESS";
export const AUHTENTICATE_USER_FAILURE = "AUHTENTICATE_USER_FAILURE";

export const FETCH_USER_START = "FETCH_USER_START";
export const FETCH_USER_SUCCESS = "FETCH_USER_SUCCESS";
export const FETCH_USER_FAILURE = "FETCH_USER_FAILURE";

export const REGISTER_USER_START = "REGISTER_USER_START";
export const REGISTER_USER_SUCCESS = "REGISTER_USER_SUCCESS";
export const REGISTER_USER_FAILURE = "REGISTER_USER_FAILURE";

export const LOGOUT_USER_START = "LOGOUT_USER_START";
export const LOGOUT_USER_SUCCESS = "LOGOUT_USER_SUCCESS";
export const LOGOUT_USER_FAILURE = "LOGOUT_USER_FAILURE";

export const CLEAR_USER_ERROR = "CLEAR_USER_ERROR";

export const SET_STAY_SIGNED_IN = "SET_STAY_SIGNED_IN";
