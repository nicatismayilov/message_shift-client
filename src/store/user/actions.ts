import { ActionTypes, Action } from "./types";
import { UserInfo } from "types/User";

// actions for fetching user data
export const fetchUserStart = (): Action => ({
	type: ActionTypes.FETCH_USER_START,
});

export const fetchUserSuccess = (userData: UserInfo): Action => ({
	type: ActionTypes.FETCH_USER_SUCCESS,
	payload: userData,
});

export const fetchUserFailure = (err: string): Action => ({
	type: ActionTypes.FETCH_USER_FAILURE,
	payload: err,
});

// action for cleaning user state
export const clearUserState = (): Action => ({
	type: ActionTypes.CLEAR_USER_STATE,
});
