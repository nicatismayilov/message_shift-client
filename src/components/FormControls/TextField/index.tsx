import React, { memo } from "react";
import classnames from "classnames";

import { inputClassFunc, labelClassFunc, hintClassFunc } from "./classnames";

import "./styles.scss";

interface Props {
	id: string;
	type: string;
	name: string;
	value: string;
	label?: string;
	error?: string;
	onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
	onBlur: (e: React.FocusEvent<HTMLInputElement>) => void;
}

const TextField: React.FC<Props> = (props) => {
	const { id, type, name, value, label, error } = props;
	const { onChange, onBlur } = props;

	const hasError = error ? true : false;
	const hasValue = value ? true : false;
	const isHintVisible = hasError;

	const inputClassName = classnames(inputClassFunc(hasError));
	const labelClassName = classnames(labelClassFunc(hasValue));
	const hintClassName = classnames(hintClassFunc(isHintVisible, hasError));

	return (
		<div className='text-field'>
			<input
				data-testid='text-field-input'
				type={type}
				name={name}
				id={id}
				autoComplete='new-password'
				className={inputClassName}
				value={value}
				onChange={onChange}
				onBlur={onBlur}
			/>

			<label data-testid='text-field-label' htmlFor={id} className={labelClassName}>
				{label}
			</label>

			<p data-testid='text-field-error' className={hintClassName}>
				{error}
			</p>
		</div>
	);
};

export default memo(TextField);
