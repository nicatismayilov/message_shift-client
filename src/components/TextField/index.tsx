import React, { useRef } from "react";
import classnames from "classnames";

import generateKey from "utils/generateKey";

import { inputClassFunc, labelClassFunc, hintClassFunc } from "./classnames";

import "./styles.scss";

interface Props {
	type: string;
	name: string;
	value: string;
	label?: string;
	error?: string;
	onChange: (value: string) => void;
}

const TextField: React.FC<Props> = (props) => {
	const { type, name, value, label, error } = props;
	const { onChange } = props;

	const id = useRef(generateKey(name));

	const hasError = error ? true : false;
	const hasValue = value ? true : false;
	const isHintVisible = hasError;

	const inputClassName = classnames(inputClassFunc(hasError));
	const labelClassName = classnames(labelClassFunc(hasValue));
	const hintClassName = classnames(hintClassFunc(isHintVisible, hasError));

	const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		const { value } = e.target;

		onChange(value);
	};

	return (
		<div className='text-field'>
			<input
				type={type}
				name={name}
				id={id.current}
				autoComplete='new-password'
				className={inputClassName}
				value={value}
				onChange={handleChange}
			/>

			<label htmlFor={id.current} className={labelClassName}>
				{label}
			</label>

			<p className={hintClassName}>{error}</p>
		</div>
	);
};

export default TextField;
