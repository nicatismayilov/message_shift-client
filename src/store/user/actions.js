import * as actionTypes from "./actionTypes";

// actions for user authentication
export const authenticateUserStart = (credentials) => ({
	type: actionTypes.AUHTENTICATE_USER_START,
	payload: credentials,
});

export const authenticateUserSuccess = (token) => ({
	type: actionTypes.AUHTENTICATE_USER_SUCCESS,
	payload: token,
});

export const authenticateUserFailure = (err) => ({
	type: actionTypes.AUHTENTICATE_USER_FAILURE,
	payload: err,
});

// actions for fetching user data
export const fetchUserStart = () => ({
	type: actionTypes.FETCH_USER_START,
});

export const fetchUserSuccess = (userData) => ({
	type: actionTypes.FETCH_USER_SUCCESS,
	payload: userData,
});

export const fetchUserFailure = (err) => ({
	type: actionTypes.FETCH_USER_FAILURE,
	payload: err,
});

// actions for user registration
export const registerUserStart = (userData) => ({
	type: actionTypes.REGISTER_USER_START,
	payload: userData,
});

export const registerUserSuccess = (token) => ({
	type: actionTypes.REGISTER_USER_SUCCESS,
	payload: token,
});

export const registerUserFailure = (err) => ({
	type: actionTypes.REGISTER_USER_FAILURE,
	payload: err,
});

// action for logout
export const logout = () => ({
	type: actionTypes.LOGOUT,
});
