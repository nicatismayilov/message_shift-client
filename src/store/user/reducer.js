import * as actionTypes from "./actionTypes";

const initialState = {
	isAuth: false,
	token: "",
	user: {},
	loading: false,
	error: "",
	staySignedIn: false,
};

const reducer = (state = initialState, action) => {
	switch (action.type) {
		case actionTypes.AUHTENTICATE_USER_START:
		case actionTypes.REGISTER_USER_START:
			return {
				...state,
				loading: true,
				isAuth: false,
			};

		case actionTypes.LOGOUT_USER_START:
			return {
				...state,
				loading: true,
			};

		case actionTypes.AUHTENTICATE_USER_SUCCESS:
		case actionTypes.REGISTER_USER_SUCCESS:
			return {
				...state,
				loading: false,
				isAuth: true,
				token: action.payload,
			};

		case actionTypes.AUHTENTICATE_USER_FAILURE:
		case actionTypes.REGISTER_USER_FAILURE:
			return {
				...state,
				error: action.payload,
				loading: false,
				isAuth: false,
				token: "",
			};

		case actionTypes.FETCH_USER_START:
			return {
				...state,
				loading: true,
			};

		case actionTypes.FETCH_USER_SUCCESS:
			return {
				...state,
				user: action.payload,
				loading: false,
			};

		case actionTypes.LOGOUT_USER_SUCCESS:
			return initialState;

		case actionTypes.FETCH_USER_FAILURE:
		case actionTypes.LOGOUT_USER_FAILURE:
			return {
				...state,
				error: action.payload,
				loading: false,
			};

		case actionTypes.CLEAR_USER_ERROR:
			return {
				...state,
				error: "",
			};

		case actionTypes.SET_STAY_SIGNED_IN:
			return {
				...state,
				staySignedIn: action.payload,
			};

		default:
			return state;
	}
};

export default reducer;
