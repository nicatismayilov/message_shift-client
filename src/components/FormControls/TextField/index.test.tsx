import React from "react";
import { render, screen } from "@testing-library/react";
import user from "@testing-library/user-event";

import TextField from "./index";

const mockHandleChange = jest.fn();
const mockHandleBlur = jest.fn();

describe("TextField", () => {
	it("should render properly", () => {
		const { baseElement } = render(
			<TextField
				type='text'
				name='text'
				value='test'
				label='text'
				error='error'
				onChange={mockHandleChange}
				onBlur={mockHandleBlur}
			/>
		);

		expect(baseElement).toMatchSnapshot();
	});

	it("should have an input type, name, label, error and value", () => {
		const mockType = "email";
		const mockName = "email";
		const mockLabel = "Test Label";
		const mockError = "Test Error";
		const mockValue = "Test Value";

		render(
			<TextField
				type={mockType}
				name={mockName}
				value={mockValue}
				label={mockLabel}
				error={mockError}
				onChange={mockHandleChange}
				onBlur={mockHandleBlur}
			/>
		);

		const Input = screen.getByTestId("text-field-input");
		expect(Input).toHaveProperty("type", mockType);
		expect(Input).toHaveProperty("name", mockName);
		expect(Input).toHaveValue(mockValue);

		const Label = screen.getByTestId("text-field-label");
		expect(Label).toHaveTextContent(mockLabel);

		const Error = screen.getByTestId("text-field-error");
		expect(Error).toHaveTextContent(mockError);
	});

	it("should call handleChange", () => {
		render(
			<TextField
				type='text'
				name='text'
				value=''
				onChange={mockHandleChange}
				onBlur={mockHandleBlur}
			/>
		);

		const Input = screen.getByTestId("text-field-input");
		user.type(Input, "test@gmail.com");
		expect(mockHandleChange).toHaveBeenCalled();
		expect(mockHandleChange).toHaveBeenCalledTimes(14);
	});
});
