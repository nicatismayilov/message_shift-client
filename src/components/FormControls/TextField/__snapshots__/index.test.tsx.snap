// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField should render properly 1`] = `
<body>
  <div>
    <div
      class="text-field"
    >
      <input
        autocomplete="new-password"
        class="text-field__input text-field__input--error"
        data-testid="text-field-input"
        id="17638775e03-text-4.ab2cb6d929934"
        name="text"
        type="text"
        value="test"
      />
      <label
        class="text-field__label text-field__label--active"
        data-testid="text-field-label"
        for="17638775e03-text-4.ab2cb6d929934"
      >
        text
      </label>
      <p
        class="text-field__hint text-field__hint--visible text-field__hint--error"
        data-testid="text-field-error"
      >
        error
      </p>
    </div>
  </div>
</body>
`;
